File: src/main/python/com/example/tutorial/dsa/medium/sorting/custom_comparator_sorter.py
"""
CustomComparatorSorter
----------------------------------
This program demonstrates custom sorting with Comparators.
The core problem solved here is Custom Sort String (LeetCode #791).

Problem Statement:
    Given a custom order string, sort another string according to this custom order.

Real UseCase:
    In a credit card offers system:
    - Sort offers by custom priority (VIP, Premium, Standard)
    - Order transactions by custom category ranking
    - Arrange merchants by partnership tier

Examples:
    - order = "cba", s = "abcd" -> Output: "cbad"
    - order = "cbafg", s = "abcd" -> Output: "cbad"

Company Tags: Facebook, Amazon

See: https://leetcode.com/problems/custom-sort-string/
"""

import sys
import os
from collections import Counter

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "..", ".."))

from com.example.tutorial.common.utils.sample_data_loader import load_offers


def custom_sort_string(order: str, s: str) -> str:
    """
    Sorts string s according to custom character order.

    LOGIC (Frequency Count + Rebuild):
        1. Count frequency of each character in s
        2. For each character in order, append it freq times
        3. Append remaining characters (not in order) at the end

    Example Walkthrough:
        order = "cba", s = "abcd"

        Count: {a:1, b:1, c:1, d:1}

        Process order "cba":
          c: append 'c' 1 time -> "c", count[c]=0
          b: append 'b' 1 time -> "cb", count[b]=0
          a: append 'a' 1 time -> "cba", count[a]=0

        Remaining (not in order):
          d: append 'd' 1 time -> "cbad"

        Result: "cbad"

    Time Complexity: O(n)
        Where n = len(s). We scan s twice (count and build).
        Like organizing books by custom shelf order - count books first,
        then place by the custom order, finally place remaining books.

    Space Complexity: O(n)
        For the result string and frequency counter.

    Args:
        order: The custom character order.
        s: The string to sort.

    Returns:
        String sorted according to custom order.
    """
    # Count frequency of each character in s.
    count = Counter(s)

    result = []

    # First, add characters in the custom order.
    for c in order:
        if c in count:
            result.append(c * count[c])
            del count[c]

    # Then, add remaining characters (not in order).
    for c, freq in count.items():
        result.append(c * freq)

    return ''.join(result)


def custom_sort_string_with_comparator(order: str, s: str) -> str:
    """
    Alternative approach using sorted() with custom key function.

    Args:
        order: The custom character order.
        s: The string to sort.

    Returns:
        String sorted according to custom order.
    """
    # Build order map: character -> priority index.
    order_map = {c: i for i, c in enumerate(order)}

    # Sort with custom key (characters not in order go to end).
    sorted_chars = sorted(s, key=lambda c: order_map.get(c, len(order)))

    return ''.join(sorted_chars)


def main():
    """Main function to demonstrate the CustomComparatorSorter."""
    print("=== CustomComparatorSorter: Custom Sorting Demo ===\n")

    # Load credit card offers from sample data
    offers = load_offers()
    print(f"Loaded {len(offers)} offers from sample data.\n")

    # Demonstrate custom string sorting
    print("--- Custom String Sort ---\n")
    order = "cba"
    s = "abcd"
    print(f"Custom order: \"{order}\"")
    print(f"Input string: \"{s}\"")
    print(f"Sorted: \"{custom_sort_string(order, s)}\"\n")

    # Demonstrate sorting objects with custom comparators
    print("--- Sorting Offers by Custom Priority ---\n")

    # Create sample offers with different priorities
    offer_data = [
        ("Offer-A", "Standard"),
        ("Offer-B", "VIP"),
        ("Offer-C", "Premium"),
        ("Offer-D", "VIP"),
        ("Offer-E", "Standard")
    ]

    # Custom priority order: VIP > Premium > Standard
    priority_order = ["VIP", "Premium", "Standard"]

    print("Original order:")
    for offer in offer_data:
        print(f"  {offer[0]} ({offer[1]})")

    # Sort by custom priority
    sorted_offers = sorted(offer_data, key=lambda x: priority_order.index(x[1]))

    print("\nSorted by priority (VIP > Premium > Standard):")
    for offer in sorted_offers:
        print(f"  {offer[0]} ({offer[1]})")

    # Additional test cases
    print("\n--- Additional Examples ---\n")
    test_cases = [
        ("cba", "abcd"),
        ("cbafg", "abcd"),
        ("xyz", "abcdef")
    ]

    for order, s in test_cases:
        result = custom_sort_string(order, s)
        print(f"order=\"{order}\", s=\"{s}\" -> \"{result}\"")


if __name__ == "__main__":
    main()

File: src/main/java/com/example/tutorial/dsa/medium/sorting/CustomComparatorSorter.java
package com.example.tutorial.dsa.medium.sorting;

import com.example.tutorial.common.datamodel.Offer;
import com.example.tutorial.common.utils.SampleDataLoader;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.stereotype.Component;

import java.util.*;

/**
 * CustomComparatorSorter
 * ----------------------------------
 * <p>This program demonstrates custom sorting with Comparators.
 * The core problem solved here is Custom Sort String (LeetCode #791).
 *
 * <p><b>Problem Statement:</b>
 * Given a custom order string, sort another string according to this custom order.
 *
 * <p><b>Real UseCase:</b> In a credit card offers system:
 * <ul>
 *   <li>Sort offers by custom priority (VIP, Premium, Standard)</li>
 *   <li>Order transactions by custom category ranking</li>
 *   <li>Arrange merchants by partnership tier</li>
 * </ul>
 *
 * <p><b>Examples:</b>
 * <ul>
 *   <li>order = "cba", s = "abcd" → Output: "cbad"</li>
 *   <li>order = "cbafg", s = "abcd" → Output: "cbad"</li>
 * </ul>
 *
 * <p><b>Company Tags:</b> Facebook, Amazon
 *
 * @see <a href="https://leetcode.com/problems/custom-sort-string/">LeetCode 791 - Custom Sort String</a>
 */
@Component
public class CustomComparatorSorter implements CommandLineRunner {
  public static void main(String[] args) {
    SpringApplication.run(CustomComparatorSorter.class, args);
  }

  @Override
  public void run(String... args) throws Exception {
    System.out.println("=== CustomComparatorSorter: Custom Sorting Demo ===\n");

    // Load credit card offers from sample data
    List<Offer> offers = SampleDataLoader.OFFERS_DTO.get();
    System.out.println("Loaded " + offers.size() + " offers from sample data.\n");

    // Demonstrate custom string sorting
    System.out.println("--- Custom String Sort ---\n");
    String order = "cba";
    String str = "abcd";
    System.out.println("Custom order: \"" + order + "\"");
    System.out.println("Input string: \"" + str + "\"");
    System.out.println("Sorted: \"" + customSortString(order, str) + "\"\n");

    // Demonstrate sorting objects with custom comparators
    System.out.println("--- Sorting Offers by Custom Priority ---\n");

    // Create sample offers with different priorities
    List<String[]> offerData = Arrays.asList(
        new String[] {"Offer-A", "Standard"},
        new String[] {"Offer-B", "VIP"},
        new String[] {"Offer-C", "Premium"},
        new String[] {"Offer-D", "VIP"},
        new String[] {"Offer-E", "Standard"}
    );

    // Custom priority order: VIP > Premium > Standard
    List<String> priorityOrder = Arrays.asList("VIP", "Premium", "Standard");

    System.out.println("Original order:");
    for (String[] offer : offerData) {
      System.out.println("  " + offer[0] + " (" + offer[1] + ")");
    }

    // Sort by custom priority
    offerData.sort((a, b) -> {
      int indexA = priorityOrder.indexOf(a[1]);
      int indexB = priorityOrder.indexOf(b[1]);
      return Integer.compare(indexA, indexB);
    });

    System.out.println("\nSorted by priority (VIP > Premium > Standard):");
    for (String[] offer : offerData) {
      System.out.println("  " + offer[0] + " (" + offer[1] + ")");
    }

    // Additional test cases
    System.out.println("\n--- Additional Examples ---\n");
    String[][] testCases = {
        {"cba", "abcd"},
        {"cbafg", "abcd"},
        {"xyz", "abcdef"}
    };

    for (String[] test : testCases) {
      String result = customSortString(test[0], test[1]);
      System.out.println("order=\"" + test[0] + "\", s=\"" + test[1] + "\" → \"" + result + "\"");
    }
  }

  /**
   * Sorts string s according to custom character order.
   *
   * <p><b>LOGIC (Frequency Count + Rebuild):</b>
   * <ol>
   *   <li>Count frequency of each character in s</li>
   *   <li>For each character in order, append it freq times</li>
   *   <li>Append remaining characters (not in order) at the end</li>
   * </ol>
   *
   * <p><b>Example Walkthrough:</b>
   * <pre>
   * order = "cba", s = "abcd"
   *
   * Count: {a:1, b:1, c:1, d:1}
   *
   * Process order "cba":
   *   c: append 'c' 1 time → "c", count[c]=0
   *   b: append 'b' 1 time → "cb", count[b]=0
   *   a: append 'a' 1 time → "cba", count[a]=0
   *
   * Remaining (not in order):
   *   d: append 'd' 1 time → "cbad"
   *
   * Result: "cbad"
   * </pre>
   *
   * <p><b>Time Complexity: O(n)</b>
   * <br>Where n = s.length. We scan s twice (count and build).
   * <br><i>Like organizing books by custom shelf order - count books first,
   * then place by the custom order, finally place remaining books.</i>
   *
   * <p><b>Space Complexity: O(n)</b>
   * <br>For the result string and frequency array (26 chars = O(1) extra).
   *
   * @param order the custom character order
   * @param s the string to sort
   * @return string sorted according to custom order
   */
  public static String customSortString(String order, String s) {
    // Count frequency of each character in s.
    int[] count = new int[26];
    for (char c : s.toCharArray()) {
      count[c - 'a']++;
    }

    StringBuilder result = new StringBuilder();

    // First, add characters in the custom order.
    for (char c : order.toCharArray()) {
      while (count[c - 'a'] > 0) {
        result.append(c);
        count[c - 'a']--;
      }
    }

    // Then, add remaining characters (not in order).
    for (int i = 0; i < 26; i++) {
      while (count[i] > 0) {
        result.append((char) ('a' + i));
        count[i]--;
      }
    }

    return result.toString();
  }

  /**
   * Alternative approach using Comparator with custom ordering.
   *
   * @param order the custom character order
   * @param s the string to sort
   * @return string sorted according to custom order
   */
  public static String customSortStringWithComparator(String order, String s) {
    // Build order map: character → priority index.
    Map<Character, Integer> orderMap = new HashMap<>();
    for (int i = 0; i < order.length(); i++) {
      orderMap.put(order.charAt(i), i);
    }

    // Convert to Character array for sorting.
    Character[] chars = new Character[s.length()];
    for (int i = 0; i < s.length(); i++) {
      chars[i] = s.charAt(i);
    }

    // Sort with custom comparator.
    Arrays.sort(chars, (a, b) -> {
      int indexA = orderMap.getOrDefault(a, 26);  // Default to end.
      int indexB = orderMap.getOrDefault(b, 26);
      return Integer.compare(indexA, indexB);
    });

    // Build result string.
    StringBuilder result = new StringBuilder();
    for (char c : chars) {
      result.append(c);
    }
    return result.toString();
  }
}
